pkgs/development/compilers/tinygo/default.nix

```
@@ -1,64 +1,74 @@
-{ lib, buildGoModule, fetchFromGitHub, llvm, clang-unwrapped, lld, avrgcc
-, avrdude, openocd, gcc-arm-embedded, makeWrapper, fetchurl }:
+{ lib
+, buildGoModule
+, fetchFromGitHub
+, makeWrapper
+, llvm
+, clang-unwrapped
+, lld
+, avrgcc
+, go
+, gcc-arm-embedded
+, avrdude
+, openocd
+}:
 
-let main = ./main.go;
-    gomod = ./go.mod;
-in
-buildGoModule rec {
+let
+  llvmMajor = lib.versions.major llvm.version;
+in buildGoModule rec {
   pname = "tinygo";
-  version = "0.16.0";
+  version = "0.20.0";
 
   src = fetchFromGitHub {
     owner = "tinygo-org";
-    repo = "tinygo";
+    repo = pname;
     rev = "v${version}";
-    sha256 = "063aszbsnr0myq56kms1slmrfs7m4nmg0zgh2p66lxdsifrfly7j";
+    sha256 = "sha256-eq1rWUVpEFRKBXbVOtYBMMYj9IMD/LeQxV7o/V2LT1s=";
     fetchSubmodules = true;
   };
 
-  overrideModAttrs = (_: {
-      patches = [];
-      preBuild = ''
-      rm -rf *
-      cp ${main} main.go
-      cp ${gomod} go.mod
-      chmod +w go.mod
-      '';
-  });
+  vendorSha256 = "sha256-K7snifSUowgpPJfQTd4nOGWLKyw+a4o0axe0tsxpzZ8=";
 
-  preBuild = "cp ${gomod} go.mod";
-
-  postBuild = "make gen-device";
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ llvm clang-unwrapped ];
 
-  vendorSha256 = "12k2gin0v7aqz5543m12yhifc0xsz26qyqra5l4c68xizvzcvkxb";
+  subPackages = [ "." ];
 
   doCheck = false;
 
+  # We need to set GOROOT to import and compile
+  # standard go libraries with tinygo
+  allowGoReference = true;
+
   prePatch = ''
```

```suggestion
  postPatch = ''
```
otherwise we break patches touching the same area of code.